registry/export.go

```diff
@@ -15,239 +18,246 @@ import (
+	if attachments == nil {
+		return nil
+	}
 
-		var v map[string]interface{}
-		if err = rows.ScanDoc(&v); err != nil {
+	prefix = path.Join(prefix, id)
+	fmt.Printf("    Exporting %s attachments\n", id)
+	for name := range attachments {
+		if err := exportCouchAttachment(writer, prefix, db, id, name); err != nil {
 			return err
 		}
+	}
 
-		if attachments, ok := v["_attachments"].(map[string]interface{}); ok {
-			for name := range attachments {
-				atts = append(atts, &attachment{docID: rows.ID(), name: name})
-			}
-			delete(v, "_attachments")
-		}
+	return nil
+}
 
-		delete(v, "_rev")
+func exportCouchDocument(writer *tar.Writer, prefix string, db *kivik.DB, rows *kivik.Rows) error {
+	id := rows.ID()
+	if strings.HasPrefix(id, "_design") {
+		return nil
+	}
 
-		var data []byte
-		data, err = json.Marshal(v)
-		if err != nil {
-			return err
-		}
+	file := path.Join(prefix, fmt.Sprintf("%s%s", id, documentSuffix))
+	fmt.Printf("    Exporting %s\n", id)
 
-		hdr := &tar.Header{
-			Name:     path.Join(dbName, rows.ID()),
-			Size:     int64(len(data)),
-			Mode:     0640,
-			Typeflag: tar.TypeReg,
-		}
-		if err = tw.WriteHeader(hdr); err != nil {
-			return err
-		}
+	var value map[string]interface{}
+	if err := rows.ScanDoc(&value); err != nil {
+		return err
+	}
 
-		fmt.Printf(`Writing document "%s/%s"... `, db.Name(), rows.ID())
-		_, err = io.Copy(tw, bytes.NewReader(data))
-		if err != nil {
-			return err
-		}
-		fmt.Println("ok.")
+	delete(value, "_rev")
+	attachments, _ := value["_attachments"].(map[string]interface{})
+	delete(value, "_attachments")
+
+	var data []byte
+	data, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+	if err = writeFile(writer, file, data, nil); err != nil {
+		return err
 	}
 
-	for _, att := range atts {
-		fmt.Printf(`Writing attachment "%s/%s/%s"... `, db.Name(), att.docID, att.name)
-		if err = writeAttachment(db, tw, dbName, att.docID, att.name); err != nil {
-			return fmt.Errorf(`Could not write attachment "%s/%s/%s": %s`, db.Name(), att.docID, att.name, err)
-		}
-		fmt.Println("ok.")
+	// Only export after document, else import will fail
+	if err := exportCouchAttachments(writer, prefix, db, id, attachments); err != nil {
+		return err
 	}
 
 	return nil
 }
 
-func writeAttachment(db *kivik.DB, tw *tar.Writer, dbName, docID, filename string) error {
-	att, err := db.GetAttachment(ctx, docID, filename)
-	if err != nil {
-		return err
-	}
-	defer att.Content.Close()
+func exportCouchDb(writer *tar.Writer, prefix string, db *kivik.DB) error {
+	name := db.Name()
+	prefix = path.Join(prefix, name)
+	fmt.Printf("  Exporting %s\n", name)
 
-	var body io.Reader
-	size := att.Size
-	if size < 0 {
-		var b []byte
-		b, err = ioutil.ReadAll(att.Content)
+	skip := 0
+	for {
+		rows, err := db.AllDocs(ctx, map[string]interface{}{
+			"include_docs": true,
+			"limit":        1000,
+			"skip":         skip,
```

Comment:
You have marked this issue as resolved, but I think it is still relevant. In particular, you have said that the CouchDB part of the export was slow, and I think it could be the reason to that.